### PR TITLE
docs(rfcs): Simple component implementation

### PR DESCRIPTION
**Pull request by sopranopillow**
# Problem statement

Currently we don't have clear guidance on how to implement or if to implement simple components that could be implemented by the user with already created components. This RFC will give guidance on what to do when we have such components and asks for feedback on which option to follow.

---

[Preview](https://github.com/microsoft/fluentui/blob/f2a0fbbdd9e13cab1a61350303ad8d8adb117094/rfcs/react-components/components/simple-component-implementation.md)

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/223571d0a6317630a2c376d278c3e4a60c10e97a'>223571d0a6317630a2c376d278c3e4a60c10e97a</a></sub>
<div id="bundleSizeComment9423"/>

File: rfcs/react-components/components/simple-component-implementation.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+```
+
+#### Pros
+
+- No need for a new package in our repo.
+- In the case of a component like persona where its design requires us to change the icon's size, we would provide best practices section that could include a size table for reference.
+
+#### Cons
+
+- Allows for issues where
```
Comment (khmakoto) [inline rfcs/react-components/components/simple-component-implementation.md:65]: Seems like this sentence is incomplete.

File: rfcs/react-components/components/simple-component-implementation.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  );
+};
+```
+
+#### Pros
+
+- We can add more design guidance
+- While the user still has to do more work than option 2, it's way easier than option 1.
+- Simpler to maintain than option 2.
+- Can make option1 easier to follow.
```
Comment (khmakoto) [inline rfcs/react-components/components/simple-component-implementation.md:114]:
```suggestion
- Can make option 1 easier to follow.
```

**Review by khmakoto — COMMENTED** [review]
My first thought is that I like the recipe approach of option 3. However, if we go that route I think we need to think more thoroughly about what the API would look like, maybe using that media object example.

File: rfcs/react-components/components/simple-component-implementation.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+As we start working on new components, there might be some components that are simple and could be built with a simple layout and the already finished components. When creating these components, we currently don't have a guidance on whether to let the user build the component or us provide the component. This RFC offers three options, one that will require the user to build the components using instructions (recipes), another where we will provide the implementation of the component, and a combination of both option 1 and 2 providing a partial implementation and a recipe.
+
+#### What makes a component a candidate for this RFC's solution?
+
+- Its implementation is simple.
+- The desired design is achievable adjusting the props of the inner components.
+- It doesn't have complex accessibility requirements.
+- It doesn't have strict design requirements.
+
```
Comment (GeoffCoxMSFT) [inline rfcs/react-components/components/simple-component-implementation.md:29]: Consider detailing the core considerations you think the team should use to make a decision of one approach vs. another.  Here's some examples:

* The decision should encourage components that provide high value beyond general purpose layout.  

For example, v9 deprecated and did not reimplement Stack.  Stack provides low value over using flexbox directly and was very general purpose.

* The decision should encourage components that are general purpose enough to be core to Fluent

Components like Field provide label, error message, and input component layout and coordination that can be used in many form scenarios. Components like a SharePoint file-list specific Card might not be reusable by others easily and maybe should live outside the core of Fluent and be implemented by the team that uses them.  This brings up a secondary question of should our decision favor a pattern that other developers building on top of fluent should use?

* The decision should prefer a solution that builds on top of our composition model.

We have a composition models that leverages slots, hooks, and context to make customization and replacement possible.  Which solutions leverage the existing composition model the best to maintain developer productivity and extensibility.

File: rfcs/react-components/components/simple-component-implementation.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+## Problem statement
+
+As we start working on new components, there might be some components that are simple and could be built with a simple layout and the already finished components. When creating these components, we currently don't have a guidance on whether to let the user build the component or us provide the component. This RFC offers three options, one that will require the user to build the components using instructions (recipes), another where we will provide the implementation of the component, and a combination of both option 1 and 2 providing a partial implementation and a recipe.
+
+#### What makes a component a candidate for this RFC's solution?
+
+- Its implementation is simple.
+- The desired design is achievable adjusting the props of the inner components.
+- It doesn't have complex accessibility requirements.
```
Comment (GeoffCoxMSFT) [inline rfcs/react-components/components/simple-component-implementation.md:27]: Consider if providing accessibility in complex scenarios is actually a great value a layout component can provide.  We might need to think about patterns and infrastructure to help coordinate interaction across components composed into a layout.

File: rfcs/react-components/components/simple-component-implementation.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+- It doesn't have strict design requirements.
+
+## Detailed Design or Proposal
+
+### Option 1: Recipes section in docsite
```
Comment (GeoffCoxMSFT) [inline rfcs/react-components/components/simple-component-implementation.md:32]: I love the idea of recipes.  I have found that the migration shims didn't work very well as recipes because people want to keep up to date with the latest and best version.  A recipe should have lots of stuff left as work for the reader to fill in.  If we find that people are just cut-and-pasting the code as-is everywhere, then it should probably be a component.

File: rfcs/react-components/components/simple-component-implementation.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+In this option we will add a section to our docs detailing how to create the component with our already created components.
+
+A great example of a recipe is the [Media Objects recipe](https://developer.mozilla.org/en-US/docs/Web/CSS/Layout_cookbook/Media_objects) by MDN.
+
+#### Folder structure inside docs
```
Comment (GeoffCoxMSFT) [inline rfcs/react-components/components/simple-component-implementation.md:38]: Storybook doesn't make recipes easy to show as the code-behind is limited to what is within the story.  Maybe we need to figure out ways to better show recipe code in storybook or find a better way to publish them.

File: rfcs/react-components/components/simple-component-implementation.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+#### Pros
+
+- We can ensure the design follows the design spec.
+- We control the accessibility behavior.
+- Makes it easier for the user.
+
+#### Cons
+
+- Another component package for a simple component.
+- More code to maintain.
```
Comment (GeoffCoxMSFT) [inline rfcs/react-components/components/simple-component-implementation.md:82]: Also, more design figmas required.

File: rfcs/react-components/components/simple-component-implementation.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+- Could be overwhelming for the user to follow the recipe.
+
+### Option 2: We provide the component
+
+For this option, we would implement the component and provide it like any other component.
+
+#### Pros
+
+- We can ensure the design follows the design spec.
+- We control the accessibility behavior.
```
Comment (ling1726) [inline rfcs/react-components/components/simple-component-implementation.md:76]: if we are going to implement a layout component we should not be dealing with accessibility, that's up to users to determine based on their use case

File: rfcs/react-components/components/simple-component-implementation.md (hunk 7 above)
Comment (ling1726) [reply rfcs/react-components/components/simple-component-implementation.md:82]: why would there be more design figmas required? I would have though that we should have one design figma to design the media layout

File: rfcs/react-components/components/simple-component-implementation.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+#### Cons
+
+- Allows for issues where the user-created component doesn't look like the design.
+- Could increase a11y issues in user-created components.
+- Could be overwhelming for the user to follow the recipe.
+
+### Option 2: We provide the component
+
+For this option, we would implement the component and provide it like any other component.
```
Comment (ling1726) [inline rfcs/react-components/components/simple-component-implementation.md:71]: just to clarify, do you mean that we would implement `Persona` or a generic `Media` layout component ?

File: rfcs/react-components/components/simple-component-implementation.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+import { Text } from '@fluentui/react-text';
+import { Image } from '@fluentui/react-image';
+import { useMediaObjectClassNames } from '@fluentui/react-recipe-hooks';
+
+const MyMediaObject = () => {
+  const mediaObjectClassNames = useMediaObjectClassNames({ labelPosition: 'below' });
+
+  return (
+    <div className={mediaObjectClassNames.mediaObject}>
+      <Image
```
Comment (ling1726) [inline rfcs/react-components/components/simple-component-implementation.md:98]: Design can sometimes have specific requirements on the inner components (i.e. Avatar), how do we handle such cases with a style hook based solution ?

File: rfcs/react-components/components/simple-component-implementation.md (hunk 9 above)
Comment (sopranopillow) [reply rfcs/react-components/components/simple-component-implementation.md:71]: If we go with this option, we will provide a Media Layout component and a Persona component.

File: rfcs/react-components/components/simple-component-implementation.md (hunk 7 above)
Comment (sopranopillow) [reply rfcs/react-components/components/simple-component-implementation.md:82]: In this case, the Persona figma would need to be broken down into two figmas (or maybe even leave it as is and say that its implementation is broken down into two components). Persona only deals with Avatar and PresenceBadge, while Media Layout would deal with any kind of media.

File: rfcs/react-components/components/simple-component-implementation.md (hunk 4 above)
Comment (sopranopillow) [reply rfcs/react-components/components/simple-component-implementation.md:27]: After thinking about it, I'll remove this requirement. I think maybe this should be left up to the engineer, there might be some scenarios in which providing some (maybe not super detailed) guidance is worthwhile.

File: rfcs/react-components/components/simple-component-implementation.md (hunk 10 above)
Comment (sopranopillow) [reply rfcs/react-components/components/simple-component-implementation.md:98]: This would be an example a part of the recipe we leave up to the user. While option 3 gives a hook, there might be some things we cannot provide, and this could be confusing.

File: rfcs/react-components/components/simple-component-implementation.md (hunk 9 above)
Comment (ling1726) [reply rfcs/react-components/components/simple-component-implementation.md:71]: Have we checked with design if there is an appetite for a reusable media layout on their side? It would requirement commitmment from both us (to implement and maintain it) and design (to design with it in mind)

File: rfcs/react-components/components/simple-component-implementation.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+- We can add more design guidance.
+- While the user still has to do more work than option 2, it's way easier than option 1.
+- Simpler to maintain than option 2.
+- Can make option 1 easier to follow.
+
+#### Cons
+
+- More code to maintain.
+- Could bring more issues by being hard to follow.
```
Comment (layershifter) [inline rfcs/react-components/components/simple-component-implementation.md:121]: You mentioned accessibility issues for Option #1, but I think it's still an issue for this option as it provides only styling.

**Comment by layershifter** [comment]
Just to clarify: what problem we are trying to solve? Create/maintain less components components that are _easy_ to build?

---

I think that recipes is a good idea as it enhances the idea of prototypes in v0 (https://fluentsite.z22.web.core.windows.net/0.64.0/prototype-text-area-autosize): we show our vision and direction there for partners.

Did you consider to add forth option with contrib package? For example, if we want to maintain less - we can create components, but encourage partners to maintain these additional bits.

**Review by marcosmoura — APPROVED** [review]

**Review by spmonahan — APPROVED** [review]

**Comment by Hotell** [comment]
> Just to clarify: what problem we are trying to solve? Create/maintain less components components that are _easy_ to build?
> 
> I think that recipes is a good idea as it enhances the idea of prototypes in v0 (https://fluentsite.z22.web.core.windows.net/0.64.0/prototype-text-area-autosize): we show our vision and direction there for partners.
> 
> Did you consider to add forth option with contrib package? For example, if we want to maintain less - we can create components, but encourage partners to maintain these additional bits.

pretty late to the party as I was not in the loop.  this is a great suggestion @layershifter  - any kind of additional components created by us will add additional maintenance for the core team.  this more "complex" scenarios should be driven by community/partners as they understand the best their needs. Core team can be leveraged as SME and help out partners achieve what they wanted in possible re-usable manner. 

the approach It could look like this:
- we have `@fluentui/labs` 
- if a component is mature enough it can be propagated to stable package, something like `@fluentui/react-component-x` ( more complex components outside the core scope )
